book/src/rfc/const_generic.md

```diff
@@ -0,0 +1,265 @@
+## Summary
+Enable const generic to improve modularization and reusability of the code.
+
+## Motivation
+In the current implementation, array initialization requires explicitly stating all element values, which becomes impractical for large arrays. Furthermore, the lack of support for constant variables as array sizes limits the ability to write functions that generically handle arrays of varying sizes. This also necessitates hardcoding loop ranges, reducing flexibility.
+
+Enabling const generic will allow define array sizes with constant variables. This would enhance function reusability and significantly improve modularization.
+
+## Detailed design
+
+### Declare a const argument
+Enable functions to define a constant argument, which can be used to define arrays with a default element or as an array size of the return type:
+
+```rust
+fn const_generic(const cst: Field) -> [Field; cst] {
+    let mut xx = [1; cst];
+    for ii in 1..cst {
+        xx[ii] = 2;
+    }
+    return xx;
+}
+```
+
+### Applying a const as a argument
+Use an existing constant to set array size directly in function parameters:
+
+```rust
+const cst = 300;
+fn const_generic(yy: Field) -> [Field; cst] {
+    let xx = [1; cst];
+    return xx;
+}
+```
+
+### Builtin functions
+
+Allow builtin functions to be defined with const arguments:
+
+```rust
+use std::to_bits;
+use std::from_bits;
+
+const num_bits = 8;
+
+fn main() {
+    let val1 = 101;
+    let bits = to_bits(num_bits, val); // bits is array ComputedExpr with len
+    let val2 = from_bits(bits); // from_bits builtin impl knows the size of the array
+    assert_eq!(val1, val2);
+}
+```
+
+The `to_bits` builtin function accepts a const variable to determine the size of the return array.
+
+```rust
+fn to_bits(const num_bits: Field, val: Field) -> [Field; num_bits]
+```
+
+The from_bits builtin function can determine the number of bits to convert based on the information that the `ComputedExprKind::Array` embeds in the `bits` variable. We will explain on `ComputedExpr` at the end to show how to improve the structure of the inputs and outputs of these builtin functions.
+
+## Type checking const variable
+
+The existing type checker is designed to do local type checks. For the cases of const variable as array size, the type checker has to trace the values globally. This is out of the scope for the design of the current type checker.
+
+Instead of checking the const variable in the type checker, we can check it in the circuit generation phase. 
```

Comment:
we never properly defined what a "const variable" was in this document BTW. Would be good to add something to the intro. Perhaps "`const` is a keyword that can be used in different places. For example, when used in front of an argument name, in a function signature, it dictates that the function must be called with a value decided at compile time (for example, with a literal, or a global constant)"

Reply:
added it to the glossary section